Color picker – Add full gradient support

barklund opened the pull request:
Full gradient support added in this PR.

With full test coverage for color picker, color input and all color-related util functions!

See it in action here: https://youtu.be/V3qsoONXUwk

Fully fixes #263.

Commit:
WIP gradient picker

Commit:
Merge branch 'master' into feature/263-gradient-picker

Commit:
Added types for color and converted color picker to use new type format

Commit:
Added CSS generator and unit tests for same

Commit:
Merge branch 'master' into feature/263-gradient-picker

Commit:
Converted color picker to new format and added migrations

Commit:
WIP update color editing component

Commit:
Now displaying color picker next to design panel

Commit:
Fix color input for texts

Commit:
Fix color for square too, plus fix text output

Commit:
Fix generated page background

Commit:
Merge branch 'master' into feature/263-gradient-picker

# Conflicts:
#	assets/src/edit-story/components/panels/style.js

Commit:
Fix all unit tests involving color picker and panels

Commit:
Added object-to-css convert and made generator only create object

Commit:
Added solid creator helper

Commit:
Added reducer logic to color picker

Commit:
Refactor color picker header, add reverse and rotate

Commit:
Fix gradient ends

Commit:
Fade in color picker

Commit:
Added labels to buttons in color picker

Commit:
Use better package for css transition

Commit:
Added a lot of keyboard interaction for gradient stop manipulation

Commit:
Fix some bugs and generators

Commit:
Fixeda number of bugs related to gradients

Commit:
Limit gradient stops to [0;1]

swissspidy commented on assets/src/edit-story/components/colorPicker/test/insertStop.js:
```diff
@@ -0,0 +1,99 @@
+const GRAY80 = { r: 51, g: 51, b: 51 };
+const MGRAY80 = { r: 51, g: 51, b: 51, a: 0.2 };
+
+expect.extend({ toBeDeepCloseTo });
```
Could also move this to `jest.setup.js` where other matchers are added

barklund replied:
I didn't really know if it was generic enough for that. Let's leave it here until we need it elsewhere.

barklund replied:
It's also a [tiny plugin with only 13 stars](https://github.com/maasencioh/jest-matcher-deep-close-to) (but MIT licensed), so I didn't want to put it in core.

Commit:
Update color component and make sure all tests pass

Commit:
Fix pattern generator and tests

Commit:
Fix some broken types

Commit:
Merge branch 'master' into feature/263-gradient-picker

Commit:
Generalized sidebar to app level

Commit:
Rename generatePatternCSS to generatePatternStyles

Commit:
Fix generator jsdoc

Commit:
Fix circular imports by importing directly

Commit:
Allow create solid to take strings

Commit:
Better helper util for creating solids from string

Commit:
Added `hasGradient` flag to color component (still unused)

Commit:
Fix typo

Commit:
Fix sidebar layering

Commit:
Reorganize color input component to smaller parts

Commit:
Added unit test for color input sub-components

Commit:
Added tests for two util functions

Commit:
Added test for last util function

Commit:
erge branch 'master' into feature/263-gradient-picker

Commit:
Add use-reduction as dependency

Commit:
Fix text color

Commit:
Fix missing opacity

Commit:
Memoize default color to avoid constant rerenders

Commit:
Move useReduction to general utils

Commit:
Some focus management - close colorpicker and re-focus old

Commit:
Require text color

Commit:
Fix test warnings by adding colors to test text elements

Commit:
Add docs to generatePatternStyles

Commit:
Always hide color picker on color input unmount

Commit:
Merge branch 'master' into feature/263-gradient-picker

# Conflicts:
#	assets/src/edit-story/components/library/panes/shapes/shapesPane.js
#	assets/src/edit-story/elements/shape/output.js
#	assets/src/edit-story/migration/migrate.js

Commit:
Make sure default context for sidebar provider has dummy functions

Commit:
Merge branch 'feature/263-gradient-picker' into feature/263-gradient-implementation

# Conflicts:
#	assets/src/edit-story/components/colorPicker/index.js
#	assets/src/edit-story/components/colorPicker/useColor.js
#	assets/src/edit-story/components/form/color.js
#	assets/src/edit-story/elements/shared/index.js
#	assets/src/edit-story/utils/generatePatternStyles.js
#	package.json

Commit:
Fix gradient picker post merge

Commit:
Generalized some constants and moved gradient line to independent components

Commit:
Add gradient stop with pointer

Commit:
Add gradient stop on pointer down rather than click

Commit:
Added mouse-based gradient stop move and remove

Converted `moveCurrentStopTo` to `moveCurrentStopBy` as it made both code snippets a lot simpler. Also changed both page display background and page preview to use style attribute rather than class to display background, as it can update much quicker.

barklund commented on tests/phpunit/tests/Story_Renderer.php:
```diff
@@ -76,12 +76,12 @@ public function test_replace_body_start_tag() {
 		};
 
-		add_action(	'web_stories_body_open', $function );
+		add_action( 'web_stories_body_open', $function );
```
I have no idea what these changes are doing here. But the pre-commit complained, I let it auto-fix, and here we are.

swissspidy replied:
Odd that this wasn‘t caught in my PR 🤔

But the change is fine.

Commit:
Fix editable hex input

Commit:
Fix editable hex input

Commit:
Use editable preview for opacity too

Commit:
Use editable preview for opacity too

Commit:
Merge branch 'feature/263-gradient-picker' into feature/263-gradient-implementation

swissspidy commented on assets/src/edit-story/components/colorPicker/useKeyDeleteStop.js:
```diff
@@ -0,0 +1,34 @@
+function useKeyDeleteStop(ref, onDelete) {
+  useKeyDownEffect(
+    ref,
+    ['del', 'backspace'],
```
Don't we support `delete` as a special case for this? See `keys.js`.

barklund replied:
Ah yes, fixed in bc05c519.

barklund replied:
Ah yes, fixed in bc05c519.

Commit:
Some minor visual corrections to previews

Commit:
Some minor visual corrections to previews

Commit:
Better handling of opacity in color input component

Commit:
Better handling of opacity in color input component

Commit:
Restructure useFocusOut to take a ref and deps array

Commit:
Remove debug defaults

Commit:
Merge branch 'feature/263-gradient-picker' into feature/263-gradient-implementation

Commit:
Use `delete` key shortcut

Commit:
Fix stop direction movement

barklund commented on assets/src/edit-story/components/colorPicker/useKeyMoveStop.js:
```diff
@@ -0,0 +1,27 @@
+import { useKeyDownEffect } from '../keyboard';
+
+function useKeyMoveStop(ref, onMove) {
+  useKeyDownEffect(ref, 'left', () => onMove(-0.01), [onMove]);
+  useKeyDownEffect(ref, 'right', () => onMove(0.01), [onMove]);
```
Fixing bug in 710b51c0 as [mentioned by @spacedmonkey](https://github.com/google/web-stories-wp/pull/497#issuecomment-596438886) (bug obviously introduced in [this refactor](https://github.com/google/web-stories-wp/pull/497/commits/3391b65d58ad518c7003c8a315513305a9af81d4#diff-1c08caaa370a6aa0afb916eda8c427ab))

barklund commented on assets/src/edit-story/components/colorPicker/useKeyAddStop.js:
```diff
@@ -0,0 +1,55 @@
+
+function useKeyAddStop(ref, onAdd, stops, currentStopIndex) {
+  const newStopPosition = getNewStopPosition(stops, currentStopIndex);
+  useKeyDownEffect(ref, 'enter', () => onAdd(newStopPosition), [
```
When any gradient stop is focused, `'enter'` creates a new gradient stop. Regarding where the stop is added, see source for `getNewStopPosition` above. This addresses [one of @spacedmonkey's comments made externally](https://github.com/WordPress/gutenberg/pull/17603#issuecomment-538122253).

How do we communicate this shortcut best?

Commit:
Merge branch 'master' into feature/263-gradient-picker

Commit:
Fix bug in escape handler

Commit:
Merge branch 'master' into feature/263-gradient-picker

Commit:
Fix layout in non-gradient color picker

Commit:
Fix tests and a11y for opacity preview

Commit:
Merge branch 'feature/263-gradient-picker' into feature/263-gradient-implementation

Commit:
Fix color picker layout for gradient

Commit:
Adding pattern-global alpha for gradients

Commit:
Removing option of adding  opacity to background

Commit:
Added some unit tests ofor excluding opacity

Commit:
Merge branch 'feature/263-gradient-picker' into feature/263-gradient-implementation

Commit:
Only allow gradieent types if enabled

Commit:
Merge branch 'master' into feature/263-gradient-picker

Commit:
Merge branch 'feature/263-gradient-picker' into feature/263-gradient-implementation

Commit:
Merge branch 'master' into feature/263-gradient-implementation

Commit:
Started integration test suite for color picker

barklund commented on assets/src/edit-story/components/colorPicker/test/colorPicker.js:
```diff
@@ -0,0 +1,109 @@
+      const promise = getResolvingPromise(onClose);
+      fireEvent.click(closeButton);
+      return promise;
+    });
```
This has a lot of duplicated logic, but my attempt at DRY'ing it up using `describe.each` resulted in a weird bug about duplicated renders. And also [KCD suggests not to nest anyway](https://kentcdodds.com/blog/avoid-nesting-when-youre-testing), though his examples are not at all related to this use-case.

barklund replied:
Currently [debugging what's going on with the failing `each`](https://twitter.com/barklund/status/1237245630441050112).

Commit:
Minor test changes

Commit:
Merge branch 'master' into feature/263-gradient-implementation

# Conflicts:
#	assets/src/edit-story/components/canvas/pagepreview/index.js
#	package.json

Commit:
Added a bunch of integration tests for color picker

Commit:
A lot more integration tests for color picker

Commit:
Added final unit tests and some minor tweaks to get color picker coverage to 100%

Commit:
Fix some other missing test coverage now that we have gradients

swissspidy committed:
Fix translator comments

swissspidy reviewed (approved):
Impressive work! 💪 

Edit: Just noticed [a small thing](https://github.com/google/web-stories-wp/pull/528/files#r390849753) while looking at another PR, but I'd say this should be handled in a new PR.

Let's get this beast in first!
